### PR TITLE
[shelly] Fix resource leak, BLU script installation, TRV init, NPE on IPv6 mDNS discovery

### PR DESCRIPTION
**Pull request by markus7017**
Includes the following fixes

- #15398: Discovery result now uses representation property serviceName rather than deviceName, which could case strange effects when adding things from the Inbox
- #15395: Gen2: Do not try to install BLU script on battery devices like PlusHT
- #15394: CoIoT attribute boost mode is now processed and results in a channel update
- #15393: Fix for TRV target temp not being initialized
- #15795: Fix resource leak when mins processing throws exception
- #15796: Fix NPE on IPv6 discovery result

### CODE REVIEW

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties [hunk 1]
```diff
@@ -42,7 +42,7 @@ message.event.triggered = Event triggered: {0}
 message.event.filtered = Event filtered: {0}
 message.coap.init.failed = Unable to start CoIoT: {0}
 message.discovery.disabled = Device is marked as non-discoverable, will be skipped
-message.discovery.protected = Device {0} reported 'Access defined' (missing userid/password or incorrect).
+essage.discovery.protected = Device {0} is protected and reports 'Access denied', check userId/password
```
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:45]: I assume first character was deleted by mistake?
```suggestion
message.discovery.protected = Device {0} is protected and reports 'Access denied', check userId/password
```

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml [hunk 2]
```diff
@@ -96,10 +96,6 @@
 			<description>@text/thing-type.config.shelly.deviceIp.description</description>
 			<context>network-address</context>
 		</parameter>
-		<parameter name="userId" type="text" required="false">
-			<label>@text/thing-type.config.shelly.userId.label</label>
-			<description>@text/thing-type.config.shelly.userId.description</description>
-		</parameter>
```
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:102]: Why is this removed - no user name, but still a password?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java:1]: Should the remaining usage of `request` also be removed as well as the method itself?
https://github.com/openhab/openhab-addons/blob/807a14a655c84706cf3c4524954ba9613bf0403f/bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java#L649-L653

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks! A few quick comments for now when browsing over the changes (not a full review).

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties (hunk 1 above)
Comment (lolodomo) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:45]: Looks like you suppressed the initial m from message.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml [hunk 3]
```diff
@@ -96,10 +96,6 @@
 			<description>@text/thing-type.config.shelly.deviceIp.description</description>
 			<context>network-address</context>
 		</parameter>
-		<parameter name="userId" type="text" required="false">
-			<label>@text/thing-type.config.shelly.userId.label</label>
-			<description>@text/thing-type.config.shelly.userId.description</description>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:101]: Please remove both entries from the i18n file if they are no more used.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java:1]: Was this change related to any of the issues, or just cleanup? There are some error handling differences in the first method, and especially the `innerRequest` methods are quite different. Perhaps related to no username required anymore?

Before:
https://github.com/openhab/openhab-addons/blob/807a14a655c84706cf3c4524954ba9613bf0403f/bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java#L655-L681

Now:
https://github.com/openhab/openhab-addons/blob/807a14a655c84706cf3c4524954ba9613bf0403f/bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpClient.java#L109-L137

Before:
https://github.com/openhab/openhab-addons/blob/807a14a655c84706cf3c4524954ba9613bf0403f/bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java#L683-L722

Now:
https://github.com/openhab/openhab-addons/blob/807a14a655c84706cf3c4524954ba9613bf0403f/bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpClient.java#L147-L220

**Comment by lolodomo** [comment]
Fix #15398
Fix #15395
Fix #15394
Fix #15393
Fix $15795
Fix #15796

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1HttpApi.java:1]: there a different routines handling GET vs. POST, retires, JSON mapping or not...
changes are related to the TRV changes
The logic is restructured and clean-up with the next PR including the basic auth fix. I decided to skip this for this PR to keep it smaller and avoid structural changes. I plan a separate PR for this.

I'll double check the timeout logic

**Comment by lolodomo** [comment]
Some review comments marked as resolved without any comment and/or change...

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java [hunk 4]
```diff
@@ -519,7 +519,7 @@ public static Map<String, Channel> createSensorChannels(final Thing thing, final
                     CHANNEL_SENSOR_VIBRATION);
         }
         // Create tilt for DW/DW2, for BLU DW create channel even tilt is currently not reported
-        if (sdata.accel != null || (profile.isBlu && sdata.lux != null)) {
+        if (sdata.accel != null || (profile.isBlu && profile.isDW && sdata.lux != null)) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:522]: Does it still match the comment above "for BLU DW create channel **even tilt is currently not reported**"? Does `profile.isDW` mean that it is reported?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java [hunk 5]
```diff
@@ -669,8 +669,8 @@ public ShellyChannel(ShellyTranslationProvider messages, String group, String ch
                 label = "";
             }
             description = getText(PREFIX_CHANNEL + typeId + ".description");
-            if (description.contains(PREFIX_CHANNEL)) {
-                description = "";
+            if (description.startsWith(PREFIX_CHANNEL)) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:672]: Why is this fix only needed here and not for the other similar lookups above?

**Comment by mstormi** [comment]
@markus7017 please adress review feedback, we're waiting to get this into the OH maintenance patch release asap.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:102]: correct, Gen2 devices hace a fixed user id "admin"

**Comment by markus7017** [comment]
I reviewed the comments (thanks @jbauer and all of your for excellent feedback). For me we are ready to go with the merge to 4.0 and 4.1. For 4.1 I'll continue with the other PR bringing more fixes, most important the basic auth handling for Gen1 and RFC conforming digest auth for Gen2.

@lolodomo Could you please go ahead with the merge, I'll then switch over to the other pending PR starting with a rebase. And yes, keeping upcoming PRs small trying to separate topics by smaller PRs.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java (hunk 5 above)
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:672]: Please do not mark comments as resolved without either providing feedback or a commit addressing the comment. For this comment I see neither, so I'm reopening. In line 668 (for example) you have exactly the same `contains` check. Why is it changed here, but not there?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java [hunk 6]
```diff
@@ -435,7 +435,10 @@ public static boolean updateSensors(ShellyThingInterface thingHandler, ShellySet
                     if (t.tmp != null) {
                         updated |= updateTempChannel(thingHandler, CHANNEL_GROUP_SENSOR, CHANNEL_SENSOR_TEMP,
                                 t.tmp.value, t.tmp.units);
-                        updated |= updateTempChannel(thingHandler, CHANNEL_GROUP_SENSOR, CHANNEL_CONTROL_SETTEMP,
+                        if (t.targetTemp.unit == null) {
+                            t.targetTemp.unit = t.tmp.units;
+                        }
```
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java:440]: I'm also reopening this one.

**Comment by markus7017** [comment]
@lolodomo let me know it something is missing

**Comment by markus7017** [comment]
@openhab/add-ons-maintainers Could somebody please do the merge?

**Comment by J-N-K** [comment]
I think the two open comments from @jlaur need feedback.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml (hunk 3 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:101]: Please comment before marking as resolved without changing anything.
Why the corresponding entries in the i18n file are not removed if you remove the config parameter ?

**Comment by jlaur** [comment]
> I think the two open comments from @jlaur need feedback.

Yes, please check the comments:
- https://github.com/openhab/openhab-addons/pull/15798#discussion_r1367975235
- https://github.com/openhab/openhab-addons/pull/15798#discussion_r1369097204
- https://github.com/openhab/openhab-addons/pull/15798#discussion_r1369091443

**Comment by lolodomo** [comment]
I just reviewed again your PR. No new comment from my side. But one of my review comment was closed without any feedback (I reopened it) and you have 2 review comments from @jlaur to answer.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml (hunk 3 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:101]: hmm, could it be that writing a comment and clicking Resolve doesn't save the comment? I'm sure that I provided feedback, but this would explain the confusion, because I gave feedback on all of them (like this one), but doesn't see it now. Super sorry if this was my handling fault.

Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:101]: In fact Gen1 requires userId, but Gen2 doesn't so I removed it in between in config2.xml like requested, but obviously my comment got lost.

**Comment by markus7017** [comment]
> I just reviewed again your PR. No new comment from my side. But one of my review comment was closed without any feedback (I reopened it) and you have also 2 review comments from @jlaur to answer.

@lolodomo @jlaur As I wrote above. Of course I gave feedback to all of your comments, BUT by writing the comment and clicking on Result instead clicking Comment and then clicking Resolve. It this that in this case the comment is not saved, which caused the confusion on your side incl. seeing that as root behavior from my side. If so, I caused the confusion and have to say sorry. I'll check each of the comments above this afternoon and let you know.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml (hunk 3 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:101]: Ok, I checked how it is done to understand. You made reference to existing i18n entries using `@test`.
And yes, the entries for userId are still usefull for gen1.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml (hunk 3 above)
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/config/config2.xml:101]: > hmm, could it be that writing a comment and clicking Resolve doesn't save the comment?

Yes, you have to post your comment by clicking "Comment", and then you can resolve it afterwards as an additional action.

**Comment by lolodomo** [comment]
@markus7017 answered to my remaining comment and he's right, nothing to change.
So code looks good to me.
@jlaur : I let you check your 2 remaining comments and merge if that is ok for you

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java:440]: uff, also lost the comment, stupid me

This is related to buggy JSON format. Esp. Gen1 devices have problems here down to the level of returning incorrect JSON. Depending on device type and firmware version the are various work arounds in the code like this one. units vs. units is correct, because that matches the returned format. And yes, someone the unit is not returned at all, therefore I default to C (because most users are from Germany)

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java (hunk 5 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:672]: hmm, one the one side startsWith is correct, because it's a prefix. However, as you outlined the are multiple of those compares - also for group name. I think I should change all of them.

I would like to merge it as is and then m are the changes as part of the next PR. Do you agree?

**Comment by markus7017** [comment]
I re-added the comments. From my pov https://github.com/openhab/openhab-addons/pull/15798#discussion_r1369097204 is still open, but for the moment I would keep it and do changes and testing with the next PR:

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java (hunk 5 above)
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:672]: Is the change related to one of the linked bugs? Wouldn't it be more consistent to fix all or none of them?

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java:440]: OK, if unit is actually missing in the payload, the work-around here should be okay. Thanks for confirming that.

It seemed suspicious that you have "unit" and "units" and one of them would be missing the unit, so I just wondered if it could have been a typo, and actually they would both carry unit.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java (hunk 5 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java]: ok, I removed it for this one and rebase should bring it with the next PR

Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java]: @jlaur could you please give feedback here so @lolodomo could click the merge button, would be very helpful so I could move on to the next PR, doing the rebase and we could work on the merge. In fact it is already completely reviewed, because this was is a subset. Nevertheless, there might be some details I need to fix.
